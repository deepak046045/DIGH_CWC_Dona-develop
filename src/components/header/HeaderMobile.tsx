'use client';

import Image from 'next/image';
import { Button } from '@mui/material';
import React, { useState } from 'react';
import { LdsButton, LdsIcon } from '@elilillyco/ux-lds-react';
import mockUserData from '@/mocks/mockUserData.json';
import {
  NavigationOptions,
  NavigationOption,
  gatedNavigationOptions,
  ungatedNavigationOptions,
} from '@/models/NavigationOptions';
import buttons from '@/scss/themes/buttons.module.scss';
import styles from './HeaderMobile.module.scss';
import headerSharedStyles from './HeaderShared.module.scss';

type HeaderMobileProps = {
  activeView: NavigationOption;
  isLoggedIn: boolean;
  signIn: () => void;
  signOut: () => void;
  handleNavigation: (arg0: NavigationOption) => void;
};

export default function HeaderMobile({
  activeView,
  isLoggedIn,
  signIn,
  signOut,
  handleNavigation,
}: HeaderMobileProps) {
  const [isMenuOpen, setIsMenuOpen] = useState<boolean>(false);

  const toggleMenuOpen = () => {
    setIsMenuOpen(!isMenuOpen);

    if (!isMenuOpen) {
      document.body.style.overflow = 'hidden';
    } else {
      document.body.style.overflow = '';
    }
  };

  function buildNavigationButton(
    navigationOption: NavigationOption
  ): JSX.Element {
    return (
      <Button
        key={navigationOption.title}
        variant="contained"
        onClick={() => handleNavigation(navigationOption)}
        fullWidth
        className={
          activeView === navigationOption
            ? styles.navigationButtonActive
            : styles.navigationButton
        }
      >
        {navigationOption.title}
      </Button>
    );
  }

  function buildSignInUpSection(): JSX.Element {
    if (isLoggedIn) {
      return (
        <div className={styles.signInOutSection}>
          <Button
            onClick={() => handleNavigation(NavigationOptions.MyProfile)}
            className={
              activeView === NavigationOptions.MyProfile
                ? headerSharedStyles.myProfileButtonActive
                : headerSharedStyles.myProfileButton
            }
            data-testid="my-profile-button"
          >
            <Image
              src={
                activeView === NavigationOptions.MyProfile
                  ? '/icons/profile-active-icon.svg'
                  : '/icons/profile-icon.svg'
              }
              alt="ProfileIcon"
              width="35"
              height="35"
            />
            {NavigationOptions.MyProfile.title}
          </Button>
          <LdsButton onClick={signOut} classes={`${buttons.primary}`}>
            Sign out
          </LdsButton>
        </div>
      );
    }

    return (
      <div className={styles.signInOutSection}>
        <LdsButton onClick={signIn} classes={`${buttons.secondary}`}>
          Sign in
        </LdsButton>
        <LdsButton classes={`${buttons.primary}`}>Sign up</LdsButton>
      </div>
    );
  }

  function buildNavigationButtonSelection(): JSX.Element {
    if (isLoggedIn) {
      return (
        <div className={`${styles.grayBar} ${styles.buttons}`}>
          {gatedNavigationOptions.map((navigationOption) =>
            buildNavigationButton(navigationOption)
          )}
        </div>
      );
    }

    return (
      <div className={`${styles.grayBar} ${styles.buttons}`}>
        {ungatedNavigationOptions.map((navigationOption) =>
          buildNavigationButton(navigationOption)
        )}
      </div>
    );
  }

  function buildMenu(): JSX.Element {
    return (
      <div className={styles.menuContainer} data-testid="mobile-menu">
        <div className={styles.menu}>{buildNavigationButtonSelection()}</div>
        {buildSignInUpSection()}
      </div>
    );
  }

  function buildMenuButton(): JSX.Element {
    if (isMenuOpen) {
      return (
        <Button onClick={toggleMenuOpen} data-testid="mobile-menu-button">
          <div className={styles.closeMenuButton}>
            Close
            <LdsIcon className={styles.closeIcon} name="X" />
          </div>
        </Button>
      );
    }

    return (
      <Button onClick={toggleMenuOpen} data-testid="mobile-menu-button">
        <Image
          src="/icons/menu-hamburger-icon.svg"
          alt="MenuHamburgerIcon"
          width="31"
          height="44"
        />
      </Button>
    );
  }

  return (
    <div className={headerSharedStyles.background}>
      <div className={headerSharedStyles.whiteBar}>
        <Image
          src="/images/kisunla-logo.svg"
          alt="Kisunla Logo"
          width={152}
          height={59}
          priority
        />
        <div className={styles.topRightControls}>
          {isLoggedIn && (
            <div className={headerSharedStyles.userWelcome}>
              Hi {mockUserData.firstName}
            </div>
          )}
          {buildMenuButton()}
        </div>
      </div>
      {isMenuOpen && buildMenu()}
    </div>
  );
}
