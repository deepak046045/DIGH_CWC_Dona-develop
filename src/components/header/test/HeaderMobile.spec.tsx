import { fireEvent, render, screen } from '@testing-library/react';
import {
  NavigationOptions,
  ungatedNavigationOptions,
} from '@/models/NavigationOptions';
import React from 'react';
import HeaderMobile from '../HeaderMobile';
import '@testing-library/jest-dom';

describe('Header Mobile Component', () => {
  it('should render header mobile', () => {
    const { container } = render(
      <HeaderMobile
        activeView={ungatedNavigationOptions[0]}
        isLoggedIn={false}
        signIn={() => {}}
        signOut={() => {}}
        handleNavigation={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  describe('when menu button clicked', () => {
    it('should open menu', () => {
      render(
        <HeaderMobile
          activeView={ungatedNavigationOptions[0]}
          isLoggedIn={false}
          signIn={() => {}}
          signOut={() => {}}
          handleNavigation={() => {}}
        />
      );
      const menuButton = screen.getByTestId('mobile-menu-button');
      expect(screen.queryByTestId('mobile-menu')).not.toBeInTheDocument();

      fireEvent.click(menuButton);

      expect(screen.getByTestId('mobile-menu')).toBeInTheDocument();
    });

    it('should toggle document overflow styling', () => {
      render(
        <HeaderMobile
          activeView={ungatedNavigationOptions[0]}
          isLoggedIn={false}
          signIn={() => {}}
          signOut={() => {}}
          handleNavigation={() => {}}
        />
      );
      const menuButton = screen.getByTestId('mobile-menu-button');

      // Open
      fireEvent.click(menuButton);
      expect(document.body.style.overflow).toBe('hidden');

      // Closed
      fireEvent.click(menuButton);
      expect(document.body.style.overflow).toBe('');
    });
  });

  describe('when navigation button clicked', () => {
    it('should call setActiveView', () => {
      const mockHandleNavigation = jest.fn();
      render(
        <HeaderMobile
          activeView={ungatedNavigationOptions[0]}
          isLoggedIn={false}
          signIn={() => {}}
          signOut={() => {}}
          handleNavigation={mockHandleNavigation}
        />
      );

      const menuButton = screen.getByTestId('mobile-menu-button');
      fireEvent.click(menuButton);

      const secondButton = screen.getByText(ungatedNavigationOptions[1].title);

      fireEvent.click(secondButton);

      expect(mockHandleNavigation).toHaveBeenCalledWith(
        ungatedNavigationOptions[1]
      );
    });
  });

  describe('when my profile button clicked', () => {
    it('should call handleNavigation', () => {
      const mockHandleNavigation = jest.fn();
      render(
        <HeaderMobile
          activeView={ungatedNavigationOptions[0]}
          isLoggedIn
          signIn={() => {}}
          signOut={() => {}}
          handleNavigation={mockHandleNavigation}
        />
      );

      const menuButton = screen.getByTestId('mobile-menu-button');
      fireEvent.click(menuButton);

      const myProfileButton = screen.getByText(
        NavigationOptions.MyProfile.title
      );

      fireEvent.click(myProfileButton);

      expect(mockHandleNavigation).toHaveBeenCalledWith(
        NavigationOptions.MyProfile
      );
    });

    it('should have correct class and icon', () => {
      const mockHandleNavigation = jest.fn();
      render(
        <HeaderMobile
          activeView={NavigationOptions.MyProfile}
          isLoggedIn
          signIn={() => {}}
          signOut={() => {}}
          handleNavigation={mockHandleNavigation}
        />
      );

      const menuButton = screen.getByTestId('mobile-menu-button');
      fireEvent.click(menuButton);

      const myProfileButton = screen.getByText(
        NavigationOptions.MyProfile.title
      );
      const myProfileButtonIcon: HTMLImageElement =
        screen.getByAltText('ProfileIcon');

      expect(myProfileButton).toHaveClass('myProfileButtonActive');
      expect(myProfileButtonIcon.src).toContain(
        '/icons/profile-active-icon.svg'
      );
    });
  });
});
