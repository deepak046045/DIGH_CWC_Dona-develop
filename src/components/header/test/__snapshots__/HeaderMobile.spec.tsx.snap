// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Mobile Component should render header mobile 1`] = `
<div>
  <div
    class="background"
  >
    <div
      class="whiteBar"
    >
      <img
        alt="Kisunla Logo"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="59"
        src="/images/kisunla-logo.svg"
        style="color: transparent;"
        width="152"
      />
      <div
        class="topRightControls"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          data-testid="mobile-menu-button"
          tabindex="0"
          type="button"
        >
          <img
            alt="MenuHamburgerIcon"
            data-nimg="1"
            decoding="async"
            height="44"
            loading="lazy"
            src="/icons/menu-hamburger-icon.svg"
            style="color: transparent;"
            width="31"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
